include/RAJA/IndexValue.hxx

```diff
@@ -261,34 +262,35 @@ protected:
   Index_type value;
 };
 
-/*!
- * \brief Helper class for convertIndex, since functions cannot be partially
- * specialized
- */
-template <typename TO, typename FROM>
-struct ConvertIndexHelper {
-  RAJA_HOST_DEVICE
-  RAJA_INLINE
-  static TO convert(FROM val) { return TO(*val); }
-};
-
-template <typename TO>
-struct ConvertIndexHelper<TO, Index_type> {
-  RAJA_HOST_DEVICE
-  RAJA_INLINE
-  static TO convert(Index_type val) { return TO(val); }
-};
 
 /*!
  * \brief Function provides a way to take either an int or any Index<> type, and
  * convert it to another type, possibly another Index or an int.
+ *
+ * This has been refactored to use SFINAE to differentiate between IndexValue
+ * derived types an non-IndexValue types, and return the correct type
+ * automatically
  */
+
```

Comment:
With C++17 (haha, not happening in RAJA any time soon), you could get away with a single function that uses `constexpr if` to determine between IndexValue types and non-IndexValue types. Since there's still a different implementation for accessing the value (direct access vs. dereference), that would be the only option that encapsulates it well.

Until we are at that point of adopting C++17, we are pretty much up a certain creek with a non-existent paddle.